### PR TITLE
Ensure custom modifiers are copied for types of overriding properties and events even in the case of dynamic/object mismatch.

### LINKED ISSUE #7674: Possibly incorrect logic in constructor of SourcePropertySymbol.
It looks like, in case of dynamic mismatch with the type of the overridden property, we will not enter the following 'if' and will not copy custom modifiers for the type.
```
                    if (_lazyType.Equals(overriddenPropertyType, ignoreCustomModifiersAndArraySizesAndLowerBounds: true, ignoreDynamic: false))
                    {
                        _lazyType = overriddenPropertyType;
                    }
```

This is how it is done for methods (CustomModifierUtils.CopyMethodCustomModifiers):
```
            TypeSymbol returnTypeWithCustomModifiers = constructedSourceMethod.ReturnType;
            if (returnType.Equals(returnTypeWithCustomModifiers, ignoreCustomModifiersAndArraySizesAndLowerBounds: true, ignoreDynamic: true))
            {
                returnType = CopyTypeCustomModifiers(returnTypeWithCustomModifiers, returnType, RefKind.None, destinationMethod.ContainingAssembly);
            }
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #7674.

@dotnet/roslyn-compiler Please review.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
Test prtest/lin/dbg/unit32 please

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs [hunk 1]
```diff
@@ -525,51 +532,79 @@ internal static void EncodeInternal(TypeSymbol type, int customModifiersCount, R
...
 
-            private static readonly Func<TypeSymbol, ArrayBuilder<bool>, bool, bool> s_encodeDynamicTransform = (type, transformFlagsBuilder, isNestedNamedType) =>
+            private class Transformer
             {
-                // Encode transforms flag for this type and it's custom modifiers (if any).
-                switch (type.TypeKind)
+                public static readonly Func<TypeSymbol, ArrayBuilder<bool>, bool, bool> WithCustomModifierFlags = new Transformer(addCustomModifierFlags: true).AddFlags;
+                public static readonly Func<TypeSymbol, ArrayBuilder<bool>, bool, bool> WithoutCustomModifierFlags = new Transformer(addCustomModifierFlags: false).AddFlags;
+
+
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs]: Nit: double blank line

File: src/Compilers/CSharp/Test/Symbol/Symbols/CustomModifiersTests.cs [hunk 2]
```diff
@@ -1690,5 +1690,231 @@ .maxstack  0
...
+}
+
+class CL2
+    : CL1
+{
+    public override dynamic[] Test
+    {
+        get
+        {
+            System.Console.WriteLine(""Get Overriden2"");
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/CustomModifiersTests.cs]: Type: overridden

File: src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs [hunk 3]
```diff
@@ -525,51 +532,79 @@ internal static void EncodeInternal(TypeSymbol type, int customModifiersCount, R
...
+                    type.VisitType(Transformer.WithCustomModifierFlags, transformFlagsBuilder);
+                }
+                else
+                {
+                    type.VisitType(Transformer.WithoutCustomModifierFlags, transformFlagsBuilder);
+                }
             }
 
-            private static readonly Func<TypeSymbol, ArrayBuilder<bool>, bool, bool> s_encodeDynamicTransform = (type, transformFlagsBuilder, isNestedNamedType) =>
+            private class Transformer
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs]: Is this class necessary or could the two uses be above be replaced by ```type.VisitType((type, builder, isNested) => AddFlags(type, builder, isNested, true))``` and ```type.VisitType(..., false))```, where `AddFlags` is a static method?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs]: Yes, I think this would work too. Do you prefer this approach?

Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs]: Yes, it would mean one fewer class to understand.

**Comment by AlekseyTs** [comment]
Feedback has been addressed.

**Comment by cston** [comment]
LGTM

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

**Comment by jaredpar** [comment]
:+1:

**Pull request merged**